src/tasks.ts

```diff
@@ -0,0 +1,102 @@
+import { spawnSync } from "child_process";
+import * as path from "path";
+import { Range, TextEditor } from "vscode";
+import { outputChannel, ws } from "./extension";
+import { TaskOutput } from "./taskRunner";
+import { NameToThemeNumber } from "./themePicker";
+import { util } from "./utility";
+
+/**
+ * D2Tasks - static functions to be used in tasks.  Functions need
+ * to be synchronous, and return their results in a callback.
+ */
+class D2Tasks {
+  public compile(text: string, log?: TaskOutput, err?: TaskOutput): string {
+    const layout: string = ws.get("previewLayout", "dagre");
+    const theme: string = ws.get("previewTheme", "default");
+    const sketch: boolean = ws.get("previewSketch", false);
+    const themeNumber: number = NameToThemeNumber(theme);
+    const d2Path: string = ws.get("execPath", "d2");
+
+    err?.("Starting Compile...");
+    err?.(`Layout: ${layout}  Theme: ${theme}  Sketch: ${sketch}\r\n`);
+
+    const args: string[] = [
+      `--layout=${layout}`,
+      `--theme=${themeNumber}`,
+      `--sketch=${sketch}`,
+      "-",
+    ];
+    const proc = spawnSync(d2Path, args, { input: text, encoding: "utf-8" });
+
+    /** proc.status: 0 - success
+     *  proc.status: 1 - errors
+     *  proc.pid: 0 - EXE Not Found
+     */
+    if (proc.pid === 0) {
+      util.showErrorToolsNotFound(proc.error?.message ?? "");
+    } else {
+      const error: string = proc.stderr?.toString() ?? "";
+
+      for (const msg of error.split("\n")) {
+        if (msg.length === 0) {
+          // Sometimes there is a blank line that we
+          // don't have to worry about
+          continue;
+        }
+
+        err?.(msg, true);
+      }
+    }
+
+    err?.("");
```

Comment:
do we need this/should it be in the else block above?

Reply:
Just to get the end message from the task system off of our output, it always needs this blank line.